File: dawn_ocaml/lib/veb_queue.ml
(* On RAM priority queues, Thorup '96 *)

open Base

type 'a veb_tree =
  | Leaf of 'a
  | Node of 'a veb_tree * 'a veb_tree

let veb_min_tree (t : 'a veb_tree) : 'a option =
  match t with
  | Leaf x -> Some x
  | Node (l, r) ->
     begin match veb_min_tree l, veb_min_tree r with
     | Some x, Some y -> Some (min x y)
     | Some x, None -> Some x
     | None, Some y -> Some y
     | None, None -> None
     end

let veb_insert (x : 'a) (t : 'a veb_tree) : 'a veb_tree =
  match t with
  | Leaf y -> if x < y then Node (Leaf x, Leaf y) else Node (Leaf y, Leaf x)
  | Node (l, r) ->
     if x < (veb_min_tree l).Contents.val then Node (veb_insert x l, r)
     else Node (l, veb_insert x r)

File: dawn_rust/src/veb_queue.rs
// On RAM priority queues, Thorup '96 

use std::cmp::Ordering;

enum VEBTree<T>
where
    T: Ord,
{
    Leaf(T),
    Node(Box<VEBTree<T>>, Box<VEBTree<T>>),
}

impl<T> VEBTree<T>
where
    T: Ord,
{
    fn min(&self) -> Option<&T> {
        match self {
            VEBTree::Leaf(x) => Some(x),
            VEBTree::Node(l, r) => match (l.min(), r.min()) {
                (Some(x), Some(y)) => Some(x.min(y)),
                (Some(x), None) => Some(x),
                (None, Some(y)) => Some(y),
                (None, None) => None,
            },
        }
    }

    fn insert(&mut self, x: T) {
        match self {
            VEBTree::Leaf(y) => {
                if x < *y {
                    *self = VEBTree::Node(Box::new(VEBTree::Leaf(x)), Box::new(VEBTree::Leaf(*y)));
                } else {
                    *self = VEBTree::Node(Box::new(VEBTree::Leaf(*y)), Box::new(VEBTree::Leaf(x)));
                }
            }
            VEBTree::Node(l, r) => {
                if let Some(min_l) = l.min() {
                    if x < *min_l {
                        l.insert(x);
                    } else {
                        r.insert(x);
                    }
                } else {
                    l.insert(x);
                }
            }
        }
    }
}
